lldb/tools/debugserver/source/MacOSX/arm64/DNBArchImplARM64.cpp

```diff
@@ -93,6 +93,55 @@ DNBArchMachARM64::SoftwareBreakpointOpcode(nub_size_t byte_size) {
 
 uint32_t DNBArchMachARM64::GetCPUType() { return CPU_TYPE_ARM64; }
 
+static std::once_flag g_cpu_has_sme_once;
+bool DNBArchMachARM64::CPUHasSME() {
+  static bool g_has_sme = false;
+  std::call_once(g_cpu_has_sme_once, []() {
+    int ret = 0;
+    size_t size = sizeof(ret);
+    if (sysctlbyname("hw.optional.arm.FEAT_SME", &ret, &size, NULL, 0) != -1)
+      g_has_sme = ret == 1;
+  });
+  return g_has_sme;
+}
+
+static std::once_flag g_cpu_has_sme2_once;
+bool DNBArchMachARM64::CPUHasSME2() {
+  static bool g_has_sme2 = false;
+  std::call_once(g_cpu_has_sme2_once, []() {
+    int ret = 0;
+    size_t size = sizeof(ret);
+    if (sysctlbyname("hw.optional.arm.FEAT_SME2", &ret, &size, NULL, 0) != -1)
+      g_has_sme2 = ret == 1;
+  });
+  return g_has_sme2;
+}
+
+static std::once_flag g_sme_max_svl_once;
+unsigned int DNBArchMachARM64::GetSMEMaxSVL() {
+  static unsigned int g_sme_max_svl = 0;
+  std::call_once(g_sme_max_svl_once, []() {
+    if (CPUHasSME()) {
+      unsigned int ret = 0;
+      size_t size = sizeof(ret);
+      if (sysctlbyname("hw.optional.arm.sme_max_svl_b", &ret, &size, NULL, 0) !=
+          -1)
+        g_sme_max_svl = ret;
+      else
+        g_sme_max_svl = get_svl_bytes();
+    }
+  });
+  return g_sme_max_svl;
+}
+
+// This function can only be called on systems with hw.optional.arm.FEAT_SME
+// It will return the maximum SVL length for this process.
+uint16_t __attribute__((target("sme"))) DNBArchMachARM64::get_svl_bytes(void) {
+  uint64_t ret = 0;
+  asm volatile("rdsvl	%[ret], #1" : [ret] "=r"(ret));
```

Comment:
Yes, when I added this there wasn't a way to get the maximum SVL but the kernel folks are adding `hw.optional.arm.sme_max_svl_b` sysctl.  You could imagine a scenario where we have generations of cores with a 64 byte SVL max, and then there's a new core some day with 128 byte SVL and there are programs which are hardcoded to the old length; maybe a process would be launched with a 64 byte SVL max by default.  It's possible to imagine debugserver having a lower max SVL than the inferior process which might have been launched with a "give me the actual full SVL" mode.  I should probably remove this backup method of getting the max SVL.

Reply:
That would be my concern, older debug server - newer CPU with a wider SME unit.

You could leave it in as a last resort but I wonder if it would even be understandable for users. If the kernel returns the registers in a buffer end to end, and they're 128 byte but debugserver thinks they're 64 byte. It'll likely 1. fail to read at all or 2. Show the first ZA row as the first two ZA rows.

As opposed to what might be useful in some very niche scenario, where you could see the bottom few bytes of each register.

Not that SME/SVE in its raw form is that useful to begin with.

Reply:
I had to install a newer kernel with the newly added sysctl to test & confirm that the sysctl works.  I'll remove the "read debugserver's max SVL and use it" fallback.